packages/iris-grid/src/sidebar/visibility-ordering-builder/VisibilityOrderingGroup.tsx

```diff
@@ -0,0 +1,171 @@
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { Button } from '@deephaven/components';
+import {
+  dhSquareFilled,
+  vsCheck,
+  vsChromeClose,
+  vsEdit,
+  vsPaintcan,
+  vsTrash,
+} from '@deephaven/icons';
+import type ColumnHeaderGroup from '../../ColumnHeaderGroup';
+
+interface VisibilityOrderingGroupProps {
+  group: ColumnHeaderGroup;
+  onDelete(group: ColumnHeaderGroup): void;
+  onColorChange(group: ColumnHeaderGroup, color: string | undefined): void;
+  onNameChange(group: ColumnHeaderGroup, name: string): void;
+  validateName(name: string): boolean;
+  isNew: boolean;
+}
+
+export default function VisibilityOrderingGroup(
+  props: VisibilityOrderingGroupProps
+): JSX.Element {
+  const {
+    group,
+    onDelete,
+    onColorChange,
+    onNameChange,
+    validateName,
+    isNew,
+  } = props;
+  const nameInputRef = useRef<HTMLInputElement>(null);
+  const colorInputRef = useRef<HTMLInputElement>(null);
+  const [name, setName] = useState(group.name);
+  const [isEditing, setIsEditing] = useState(isNew);
+  const isValid =
+    (isEditing && name === group.name) ||
+    (name !== group.name && validateName(name));
+
+  useEffect(
+    function focusEditInput() {
+      if (isEditing && nameInputRef.current) {
+        nameInputRef.current.select();
+      }
+    },
+    [isEditing]
+  );
+
+  if (isEditing) {
+    const handleConfirm = () => {
+      if (isValid) {
+        onNameChange(group, name);
+        setIsEditing(false);
+      }
+    };
+
+    const handleCancel = () => {
+      setName(group.name);
+      setIsEditing(false);
+    };
+
+    const handleEditKeyDown = (e: React.KeyboardEvent): void => {
+      if (e.key === 'Enter') {
+        e.stopPropagation();
+        handleConfirm();
+      }
+
+      if (e.key === ' ') {
+        e.stopPropagation();
+      }
+
+      if (e.key === 'Escape') {
+        handleCancel();
+      }
+    };
+
+    return (
+      <div className="group-editing py-1">
+        <div style={{ display: 'flex', alignItems: 'center', flexGrow: 1 }}>
+          <input
+            ref={nameInputRef}
+            className={classNames('form-control', 'mr-1', {
+              'is-invalid': !isValid,
+            })}
+            style={{ flexGrow: 1 }}
+            value={name}
+            onChange={e => setName(e.target.value)}
+            onKeyDown={handleEditKeyDown}
+          />
+          <Button
+            kind="ghost"
+            className="p-1"
+            icon={vsCheck}
+            tooltip="Confirm"
+            onClick={handleConfirm}
+          />
+          <Button
+            kind="ghost"
+            className="p-1"
+            icon={vsChromeClose}
+            tooltip="Cancel"
+            onClick={handleCancel}
+          />
+        </div>
+        {!isValid && (
+          <p className="mb-0 validate-label-error text-danger">
+            Invalid or duplicate name
+          </p>
+        )}
+      </div>
+    );
```

Comment:
Fair point. I don't think `useCallback` is necessary for the handle methods here though since they don't get passed down any further and will change on every update anyway. This only renders base elements so they'll all re-render if anything triggers a render. And controlled inputs will change the callback every time a key is pressed so memoizing them doesn't prevent re-renders